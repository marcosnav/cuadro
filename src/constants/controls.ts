/**
 * Available control types
 */
export enum ControlType {
  NEW_GAME = 'NEW_GAME',
  RESTART_GAME = 'RESTART_GAME',
  SEE_ORIGINAL = 'SEE_ORIGINAL',
}
