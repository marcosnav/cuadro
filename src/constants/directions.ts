/**
 * SwipeDirection, one of UP, RIGHT, LEFT, DOWN
 */
export enum SwipeDirection {
  UP = 'UP',
  RIGHT = 'RIGHT',
  DOWN = 'DOWN',
  LEFT = 'LEFT',
}
