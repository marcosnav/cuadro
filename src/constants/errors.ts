export enum errors {
  NO_INTERNET = 'NO_INTERNET',
  UNEXPECTED_ERROR = 'UNEXPECTED_ERROR',
  UNRESPONSIVE_IMAGE_PROVIDER = 'UNRESPONSIVE_IMAGE_PROVIDER',
}
