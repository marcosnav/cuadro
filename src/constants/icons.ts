import check from './../assets/svg/check.svg';
import image from './../assets/svg/image.svg';
import newicon from './../assets/svg/new.svg';
import refresh from './../assets/svg/refresh.svg';

export const icons: { [key: string]: string } = {
  check,
  image,
  new: newicon,
  refresh,
};
