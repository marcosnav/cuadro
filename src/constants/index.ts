export * from './status';
export * from './errors';
export * from './icons';
export * from './theme';
export * from './controls';
export * from './directions';
export * from './keycodes';
