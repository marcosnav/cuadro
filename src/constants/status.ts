export enum status {
  LOADING_IMAGE = 'LOADING_IMAGE',
  DISPLAY_ORIGINAL = 'DISPLAY_ORIGINAL',
  PLAYING_GAME  = 'PLAYING_GAME ',
  FINISHED_GAME = 'FINISHED_GAME',
}
