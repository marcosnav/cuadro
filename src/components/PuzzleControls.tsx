import React, { FC } from 'react';
import { ControlType, Status } from './../constants';
import ControlButton from './ControlButton';
import * as S from './styled';

interface IProps {
  status?: Status;
  onNewGame: () => void;
  onRestart: () => void;
  onSeeOriginal: () => void;
}


/**
 * Controls Component
 * @param props { newOnly: boolean }
 */
const PuzzleControls: FC<IProps> = ({ status, onNewGame, onRestart, onSeeOriginal }) => {
  const displayOriginal = status === Status.DISPLAY_ORIGINAL;
  const playingGame = status === Status.PLAYING_GAME;
  const startingGame = status === Status.STARTING_NEW_GAME;
  const finishedGame = status === Status.FINISHED_GAME;

  const secondaryControls = () => {
    return (
      [(
        <ControlButton
          key='restart'
          action={onRestart}
          type={ControlType.RESTART_GAME}
          disabled={displayOriginal}
        >
          Restart
        </ControlButton>
      ),
      (
        <ControlButton
          key='seeoriginal'
          action={onSeeOriginal}
          type={ControlType.SEE_ORIGINAL}
        >
          {displayOriginal ? 'Hide' : 'See'} Original
        </ControlButton>
      )]
    );
  };

  return (
    <S.Controls newOnly={startingGame || finishedGame}>
      <ControlButton
        action={onNewGame}
        type={ControlType.NEW_GAME}
        disabled={displayOriginal}
      >
        New Game
      </ControlButton>
      {playingGame || displayOriginal ? secondaryControls() : null}
    </S.Controls>
  );
};

PuzzleControls.defaultProps = {
  status: Status.PLAYING_GAME,
};

export default PuzzleControls;
