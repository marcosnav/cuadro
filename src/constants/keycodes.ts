import { SwipeDirection } from './directions';

/**
 * KeyCodes to match for keypress events
 */
export enum KeyCodes {
  DOWN = 40,
  LEFT = 37,
  RIGHT = 39,
  UP = 38,
  ENTER = 13,
}

/**
 * Map key code with swipe direction
 */
export const KeySwipeMap: { [key: number]: SwipeDirection } = {
  40: SwipeDirection.DOWN,
  37: SwipeDirection.LEFT,
  39: SwipeDirection.RIGHT,
  38: SwipeDirection.UP,
};
