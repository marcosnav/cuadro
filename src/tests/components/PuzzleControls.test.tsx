import { mount, shallow } from 'enzyme';
import React from 'react';
import PuzzleControls from './../../components/PuzzleControls';
import { Status } from './../../constants';

const mockFn = () => '';

describe('<PuzzleControls/>', () => {
  test('renders properly', () => {
    const component = shallow(
    <PuzzleControls onNewGame={mockFn} onRestart={mockFn} onSeeOriginal={mockFn} />);
    const controls = component.find('ControlButton');
    expect(controls.getElements().length).toBe(3);
    expect(component).toMatchSnapshot();
  });

  describe('with status FINISHED_GAME', () => {
    test('shows only the new game option', () => {
      const component = shallow(
      <PuzzleControls status={Status.FINISHED_GAME} onNewGame={mockFn} onRestart={mockFn} onSeeOriginal={mockFn} />);
      const controls = component.find('ControlButton');
      expect(controls.getElements().length).toBe(1);
    });
  });

  describe('with status DISPLAY_ORIGINAL', () => {
    test('new game and reload must be disabled', () => {
      const component = mount(
      <PuzzleControls status={Status.DISPLAY_ORIGINAL} onNewGame={mockFn} onRestart={mockFn} onSeeOriginal={mockFn} />);
      const controls = component.find('ControlButton');
      expect(controls.getElements().length).toBe(3);
      expect(controls.first().props().disabled).toBe(true);
      expect(controls.at(1).props().disabled).toBe(true);
      expect(controls.last().props().disabled).toBe(false);
      expect(controls.last().text()).toBe('Hide Original');
    });
  });

  describe('on actions', () => {
    test('they must be triggered', () => {
      const newGameSpy = jest.fn();
      const restartSpy = jest.fn();
      const originalSpy = jest.fn();
      const component = mount(
      <PuzzleControls onNewGame={newGameSpy} onRestart={restartSpy} onSeeOriginal={originalSpy} />);
      const controls = component.find('ControlButton');
      const newGameControl = controls.first();
      const restartControl = controls.at(1);
      const seeOriginalControl = controls.last();
      newGameControl.simulate('click');
      expect(newGameSpy).toHaveBeenCalled();
      restartControl.simulate('click');
      expect(restartSpy).toHaveBeenCalled();
      seeOriginalControl.simulate('click');
      expect(originalSpy).toHaveBeenCalled();
    });
  });
});
