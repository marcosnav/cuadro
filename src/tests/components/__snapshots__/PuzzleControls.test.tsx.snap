// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PuzzleControls/> renders properly 1`] = `
<Controls
  newOnly={false}
>
  <ControlButton
    action={[Function]}
    disabled={false}
    type="NEW_GAME"
  >
    New Game
  </ControlButton>
  <ControlButton
    action={[Function]}
    disabled={false}
    type="RESTART_GAME"
  >
    Restart
  </ControlButton>
  <ControlButton
    action={[Function]}
    disabled={false}
    type="SEE_ORIGINAL"
  >
    See
     Original
  </ControlButton>
</Controls>
`;
